File: leetcode/1022.从根到叶的二进制数之和.go
package main

/*
 * @lc app=leetcode.cn id=1022 lang=golang
 *
 * [1022] 从根到叶的二进制数之和
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start
func sumRootToLeaf(root *TreeNode) int {
	return dfs(root, 0, 0)
}

func dfs(node *TreeNode, num int, res int) int {
	if node == nil {
		return res
	}
	num = num*2 + node.Val
	if node.Left == nil && node.Right == nil {
		res += num
		return res
	}
	res = dfs(node.Left, num, res)
	res = dfs(node.Right, num, res)
	return res
}

// @lc code=end

File: leetcode/1022.从根到叶的二进制数之和.py
#
# @lc app=leetcode.cn id=1022 lang=python3
#
# [1022] 从根到叶的二进制数之和
#
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(node: TreeNode, num: int):
            nonlocal res
            if not node:
                return
            num = num * 2 + node.val
            if not (node.left or node.right):
                res += num
                return
            dfs(node.left, num)
            dfs(node.right, num)

        dfs(root, 0)

        return res

# @lc code=end
